[REEF-1160] Deprecate O.A.R.Driver.ContextConfiguration and add O.A.R…

afchung opened the pull request:
….Common.Context.ContextConfiguration

This addressed the issue by
  * Deprecated O.A.R.Driver.ContextConfiguration (and options) and add O.A.R.Common.Context.ContextConfiguration (and options).
  * Added tests for currently the only functionality of ContextConfigurations -- allowing submission of Context IDs.
  * Deprecated default context handlers and created new classes for them in O.A.R.Common.

JIRA:
  [REEF-1160](https://issues.apache.org/jira/browse/REEF-1160)

afchung commented:
@jwang98052 could you please have a look when you have time? This will be blocking REEF-1063, REEF-1062, and REEF-1144. Thanks!

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs:
```diff
@@ -0,0 +1,39 @@
+                return Optional<ContextMessage>.Empty();
+            }
+
+            set
```
Why is there an empty `set` here?

afchung replied:
`IContextMessageSource` requires both `get` and `set` to be implemented, which I think is a bad idea, but this should be addressed in a separate PR.

markusweimer replied:
Agreed. I filled [REEF-1166](https://issues.apache.org/jira/browse/REEF-1166) for it.

markusweimer replied:
That PR just went in. We should remove this setter.

afchung replied:
Sounds good. I'll make a merge commit and fix this.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs:
```diff
@@ -36,17 +38,21 @@ internal sealed class ContextManager : IDisposable
         private ContextManager(
             HeartBeatManager heartBeatManager,
-            EvaluatorSettings evaluatorSetting)
+            EvaluatorSettings evaluatorSetting,
```
typo? `evaluatorSetting` -> `evaluatorSettings`

afchung replied:
Good catch.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -93,14 +90,35 @@ internal sealed class ContextRuntime
+        /// <param name="serviceInjector"></param>
+        /// <param name="contextConfiguration">the Configuration for this context.</param>
+        /// <param name="parentContext"></param>
+        public ContextRuntime(
```
Does this have to be `public`?

afchung replied:
`ContextRuntime` is an `internal` class.

markusweimer replied:
fair enough

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Context/ContextConfiguration.cs:
```diff
@@ -30,6 +30,7 @@
 {
     [ClientSide]
+    [Obsolete("Obsoleted in 0.14, please use Org.Apache.REEF.Common.ContextConfiguration instead.")]
     public sealed class ContextConfiguration : ConfigurationModuleBuilder
```
Can you also use the new named parameters instead?

markusweimer commented:
I had another look. The implementations of contexts on the Evaluator side is a stop-gap until we do the proper rewrite, right?

afchung commented:
@markusweimer yes, this is just a stop-gap such that we support both the new and the old `ContextConfiguration` s in the meantime. The old `ContextConfiguration` is really just a container for `ContextId` at this point though...

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs:
```diff
@@ -0,0 +1,92 @@
+namespace Org.Apache.REEF.Common.Context
+{
+    [ClientSide]
+    public sealed class ContextConfiguration : ConfigurationModuleBuilder
```
You might use this opportunity to rename it as ContextConfigurationModule to distinguish it from another ContextConfiguration.

afchung replied:
I would like to keep the parity between C# and Java. Java uses `ContextConfiguration`. The other `ContextConfiguration` will eventually be deleted.

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs:
```diff
@@ -0,0 +1,59 @@
+    ///  Configuration parameters for ContextConfiguration module.
+    /// </summary>
+    [ClientSide]
+    public sealed class ContextConfigurationOptions : ConfigurationModuleBuilder
```
Why inherit from ConfigurationModuleBuilder? Should clean up.

afchung replied:
Not sure, it was there from before. I'll clean this up.

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -55,21 +55,18 @@ internal sealed class ContextRuntime
-                    new ArgumentException("contextConfiguration is not of type ContextConfiguration"), LOGGER);
-            }
-            _contextLifeCycle = new ContextLifeCycle(config.Id);
+            _contextLifeCycle = new ContextLifeCycle(id);
```
You make it injectable, but still use new here.

afchung replied:
We are having two paths, one is via injection, and one is via the to-be-removed path. This is the to-be-removed path.

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -93,14 +90,35 @@ internal sealed class ContextRuntime
             IInjector serviceInjector,
             IConfiguration contextConfiguration)
-            : this(serviceInjector, contextConfiguration, Optional<ContextRuntime>.Empty())
+            : this(id, serviceInjector, contextConfiguration, Optional<ContextRuntime>.Empty())
```
If it is going to be obsolete, why brother to change it to add id parameter here?

afchung replied:
Because we still need to support the old way to instantiate things in the meantime (i.e. using `ContextConfiguration.Id`.

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -141,8 +170,9 @@ public ContextRuntime SpawnChildContext(IConfiguration contextConfiguration, ICo
                 {
-                    IInjector childServiceInjector = _serviceInjector.ForkInjector(serviceConfiguration);
-                    var childContext = new ContextRuntime(childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(this));
+                    var childServiceInjector = _serviceInjector.ForkInjector(serviceConfiguration);
```
Was _ServiceInjector not created using serviceConfiguration?

afchung replied:
this is the child service configuration, they are different.

jwang98052 replied:
Then you might want to name serviceConfiguration as childServiceConfiguration to make it clear.

afchung replied:
Done

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -115,6 +133,17 @@ public Optional<ContextRuntime> ParentContext
+                return new ContextRuntime(childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(parentRuntime));
+            }
+            
+            return new ContextRuntime(actualContextConfiguration.Id, childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(parentRuntime));
```
still use obsoleted ContextRuntime?

afchung replied:
Yes, it will be Obsoleted in the 0.14 release, but we cannot yet remove it, so we have to support both paths prior to the 0.15 release.

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs:
```diff
@@ -35,35 +35,33 @@ internal sealed class RootContextLauncher
             if (_rootContext == null)
             {
-                _rootContext = GetRootContext(_rootServiceInjector, _rootContextConfiguration);
+                _rootContext = new ContextRuntime(Id, _rootServiceInjector, _rootContextConfiguration);
```
This is still obsolete one. Should you put TODO and add a Jira for it?

afchung replied:
Filed REEF-1167.

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSimpleContext.cs:
```diff
@@ -0,0 +1,219 @@
+
+        private static IConfiguration MergeAssembliesConfiguration(IConfiguration configuration)
+        {
+            return TangFactory.GetTang().NewConfigurationBuilder(configuration)
```
This should be not needed. You can remove it and have a try.

jwang98052 commented:
I completed my pass. 

Jira mentioned it is file moving. But looks like you have changes more than that in this PR. Could you describe it in either Jira or this PR description to make it clear.

afchung commented:
@jwang98052 @markusweimer I've addressed your comments, modified the description JIRA, and opened REEF-1167 for the removal phase. Please have another look, thanks for the reviews!

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs:
```diff
@@ -0,0 +1,59 @@
+    /// <summary>
+    ///  Configuration parameters for ContextConfiguration module.
+    /// </summary>
+    /// TODO[JIRA REEF-1167]: Remove this class.
```
Isn't it the other copy of this class we'd like to remove? Also, it would be good for old a new code paths to share their named parameters.

afchung replied:
Whoops, added to the wrong class. Fixed.

afchung commented:
@jwang98052 @markusweimer Could you please have another look when available? Thanks!

markusweimer commented:
Looks good to me. @jwang98052?

jwang98052 commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs:
```diff
@@ -276,8 +282,18 @@ private void AddContext(AddContextProto addContextProto)
+                IConfiguration contextConfiguration;
+                try
+                {
+                    contextConfiguration = _serializer.FromString(addContextProto.context_configuration);
```
In which situation, this deserialization may get exception?

jwang98052 commented:
I will test and merge

afchung commented:
@jwang98052 after reading your earlier question, I've realized a mistake I've made in the PR, but I didn't think you would merge this since I have not yet answered the question. Sorry! I've created a new item (REEF-1172) to address the mistake, and will create a new PR soon. Please have a look at it. Thanks!
